docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+
+In the *One pipeline* model, the only notable perceived difference between
+the *Active-Passive* model are the worst performances obtained in the event of failure.
+
+In this scenario, the system should pay also the time needed to fully load a new
+pipeline, that unlike that *Active-Passive* model, is not preloaded.
+
+So, supposing that the new pipeline has scheduled to load on a node where the
+s3gw image is not cached, the system should pay the time needed to download
+the image from the registry before starting it.
+
+## Investigation's Rationale
+
+The 3 models described above have different performances and different implementation
+efforts. While the *Active-Active* model is expected to require a significant
+development effort due to its inherent complex nature, the *Passive-Active* model
+seems to offer a lower degree of implementation difficulty.
+
+In particular, the *One Active pipeline* model, is expected to work with nothing
+but the primitives normally available on any Kubernetes cluster.
+
+Given that the backend s3gw architecture is composed by:
+
+- one *stateless* `radosgw` process associated with a *stateful* `PV`
+
+it is supposed that, on `radosgw` process failure, a simple reschedule
+of its POD is enough to fulfill this simple HA model.
+A clarification is needed here: the `PV` *state* is guaranteed by a third party
+service, that is: **Longhorn**.
+
+Longhorn ensures that a `PV` (along with its content) is always available on
+any cluster's node, so that, a POD can mount it regardless of its allocation on the
+cluster.
+
+Relying on this assumption, while we are still in the process to decide what model
+we want to actually engage, the *One Active pipeline* model seems for sure
+the easiest thing to explore and to investigate.
+
+## Failure cases
+
+The `PV` state is kept coherent by Longhorn, so errors at this level are assumed
+NOT possible (application level corruptions to the `PV`'s state ARE possible).
+
+What are the failure cases that can happen for s3gw?
+
+Let's examine the following scenarios:
+
+1. `radosgw`'s POD failure or rescheduling
+2. Cluster's node failure
+3. `radosgw`'s failure due to a bug not related to a certain input pattern
+4. `radosgw`'s failure due to a bug related to a certain input pattern
+5. `PV` Data corruption at application level due to `radosgw`'s bug
+6. `PV` Data corruption at application level due to `radosgw`'s anomalous exit
+
+We are supposing all these bugs or conditions to be fatal for the s3gw's process
+so that they trigger an anomalous exit.
+
+### radosgw's POD failure
+
+This case is when the `radosgw` process stops due to a failure of its POD.
+This case applies also when Kubernetes decides to reschedule the POD
+to another node, eg: when the node goes low on resources.
+
+This can be thought as an infrastructure issue independent to the s3gw.
+In this case, the *One Active pipeline* model fully restores the service by
```

Comment:
This would often be called a "switch-over" in an HA cluster - e.g., an administratively orchestrated transition of the service to a new node (say, for maintenance/upgrade/etc reasons).
This has the advantage of being schedulable, so it can happen at times of low load if these exist.

But in combination with proper interaction with the ingress - pausing requests there instead of failing them - we should be able to mask these cleanly.

Bonus: this is also what we need to seamlessly restart on upgrade/update of the pod itself transparently.

This is 100% within our scope, I'd say.